### ISSUE TITLE
Feature idea: Effect highlighting

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#4276]
I want to be able to click on an effect in the header of a function and have the expressions are cause that effect to be highlighted, like so:
![image](https://user-images.githubusercontent.com/32037926/179214501-ccd7a46c-dfde-42e0-8ca3-596bd76b46d0.png)

This could be implemented in gradually more complex steps:
1) Highlight base expressions: `ref 2 @ r`, `[] @ r`, `do F()`
2) Highlight direct function calls `f(1, 2)` where `def f(x: Int32, y: Int32): Int32 & r`
3) Highlight all non-composite expressions that cause the effect.

We've discussed implementing it by gathering all expressions that has the effect in question and the filtering out expressions that cannot have caused it. `e1 ; e2` would be filtered away fx. since it must be from `e1` or `e2`.

**Comment by magnus-madsen** [comment]
I think an easy starting point is to look at the do-effect.

**Comment by magnus-madsen** [comment]
@LoZander Would it be easy to extend the HighlightProvider with this as a proof of concept?

The simplest scenario I can think of is that I want to click on an effect and then it should highlight all uses of that effect.

**Comment by LoZander** [comment]
Yeah, I think that should be doable. I'll take a look at it once it soon :)

**Comment by LoZander** [comment]
I've started work on this :)

**Comment by LoZander** [comment]
Oh, I missed the part where both of you suggested doing this incrementally. I kinda just went all out ^^'

**Comment by LoZander** [comment]
My approach so far is to gather all expressions in the file and then depending on the type of expression, decide if we should consider highlighting it. For instance, for `2 + f(3)` for some impure `f`, we shouldn't highlight the entire expression, but just the most "precise" source which would be `f(3)`. When we "consider" the expression, we also filter it out if it's not in the function we care about.

**Comment by LoZander** [comment]
> I want to be able to click on an effect in the header of a function and have the expressions are cause that effect to be highlighted, like so: ![image](https://user-images.githubusercontent.com/32037926/179214501-ccd7a46c-dfde-42e0-8ca3-596bd76b46d0.png)
> 
> This could be implemented in gradually more complex steps:
> 
> 1. Highlight base expressions: `ref 2 @ r`, `[] @ r`, `do F()`
> 2. Highlight direct function calls `f(1, 2)` where `def f(x: Int32, y: Int32): Int32 & r`
> 3. Highlight all non-composite expressions that cause the effect.
> 
> We've discussed implementing it by gathering all expressions that has the effect in question and the filtering out expressions that cannot have caused it. `e1 ; e2` would be filtered away fx. since it must be from `e1` or `e2`.

I suppose I went directly for step 3 ^^'

**Comment by LoZander** [comment]
This first iteration just highlights op applications (`do` expressions) that perform the effect being highlighted. It only highlights these `do` expressions for the effect in question and it does so no matter what kind of occurrence of the effect in question. We simply highlight all the `do` expressions in the file that produce the effect in question.

Further work would ofc. add highlighting for more complex expressions that produce effects, but it might also be worth considering whether expressions that produce the effect should be highlighted no matter what kind of occurrence of the effect in question the cursor is on and whether we want _all_ of the sources to be highlighted. Certainly when the cursor is on the effect in the function spec, but what about when the cursor is on the definition of the effect or some other occurrence of the effect symbol? And maybe we only care about the occurrences in function in question?

**Comment by magnus-madsen** [comment]
Fixed in master.
